rfd/0080-yubikey-piv.md

```diff
@@ -0,0 +1,429 @@
+    touch_policy: cached
+---
+kind: role
+version: v5
+metadata:
+  name: piv-pin
+spec:
+  piv_attestation: 
+    mode: required
+    pin_policy: always
+---
+kind: role
+version: v5
+metadata:
+  name: piv-touch
+spec:
+  piv_attestation: 
+    mode: required
+    touch_policy: always
+```
+- Users with `piv-pin` role must use PIV login with `pin=always` and `touch=cached|always`
+- Users with `piv-touch` role must use PIV login with `pin=once|always` and `touch=always`
+- Users with both the `piv-pin` and `piv-touch` roles must use PIV login  with `pin=always` and `touch=always`
+- Users with neither the `piv-pin` nor `piv-touch` roles are not required to use PIV login
+
+#### Retrieve user's PIV attestation requirements during login
+
+Since PIN and touch policies are configured locally on a PIV private key, we cannot delegate the task of setting policies for a private key to the server. Additionally, since these policies depend on Cluster Auth Preference and Role settings, there is no way for us to give the required settings to the user before they login.
+
+To work around this limitiation, we will add another gRPC endpoint `GetPIVAssertionRequirements` for a user to fetch their PIV requirements:
+
+```
+rpc GetPIVAssertionRequirements() returns (PIVAssertionRequirements);
+
+message PIVAssertionRequirements {
+    string Mode = 1;
+    string PINPolicy = 2;
+    string TouchPolicy = 3;
+}
+```
+
+This can be performed after the initial login/signing flow, and before attestation. If the user did not use `--piv` or did not provide strict enough `--piv-pin-policy` or `-piv-touch-policy`, then the user will see a warning message, and `tsh login` will automatically try to generate a valid PIV key and re-perform the login/singing flow with the new key.
+
+### PIV secret management - Management Key, PUX, and PIN
+
+Some PIV operations require [adminstrative access](https://developers.yubico.com/PIV/Introduction/Admin_access.html), which require one or more of the following secrets:
+
+| Name           | size     | default value                                      | function                                  |
+|----------------|----------|----------------------------------------------------|-------------------------------------------|
+| Management Key | 24 bytes | `010203040506070801020304050607080102030405060708` | private key and certificate management    |
+| PIN            | 8 chars  | `123456`                                           | sign and decrypt data, reset pin          |
+| PUK            | 8 chars  | `12345678`                                         | reset PIN when blocked by failed attempts |
+
+The Management Key must be provided during login, and the PIN must be provided during login and during handshakes. Using the default secrets is not recommended, so we will prompt the user for these values when they are needed.
+
+```
+> tsh login --piv
+Enter the PIV card's PIN:
+Enter the PIV card's Management Key:
```

Comment:
This is only needed when you use PIN and only the first time, correct?

Also, if you set only the "touch" policy, then you'll just need to tap the key?

Reply:
The user always needs to provide the Management Key during login. Since we are setting the Management Key in PIN protected metadata, they would need to provide their PIN on login (instead of management key) regardless of the PIN policy.

I'll flip this prompt to ask for management key first in case they didn't set it in PIN protected metadata, in which case they wouldn't need to provide PIN at all if PIN policy is never:

```
> tsh login --piv
Enter the PIV card's Management Key [blank to retrieve it from PIN protected metadata]:
Enter the PIV card's PIN:
```

Also, PIN policy `once` means that you need to provide the PIV PIN once per `tsh` request, not just on login.